flow/record/base.py

```diff
@@ -338,243 +335,308 @@ class RecordFieldSet(list):
-                ).format(field=field.name, default=default)
-            unpack_code += "\t\t)"
-
-        init_code += "\t\t__self._generated = _generated or _utcnow()\n\t\t__self._version = RECORD_VERSION"
-        # Store the fieldtypes so we can enforce them in __setattr__()
-        field_types = "{\n"
-        for field in all_fields:
-            field_types += "\t\t{field!r}: _field_{field}.type,\n".format(field=field)
-        field_types += "\t}"
-
-        code = RECORD_CLASS_TEMPLATE.format(
-            name=name.replace("/", "_"),
-            args=args,
-            slots_tuple=tuple(all_fields.keys()),
-            init_code=init_code,
-            unpack_code=unpack_code,
-            field_types=field_types,
-        )
+        Returns:
+            Mapping of required fields
+        """
+        return OrderedDict([(k, RecordField(k, v)) for k, v in RESERVED_FIELDS.items()])
 
-        code = code.replace("\t", "    ")
-        c = compile(code, "<record code>", "exec")
+    @property
+    def fields(self) -> Mapping[str, RecordField]:
+        """
+        Get fields mapping (without required fields). eg:
 
-        data = {
-            "desc": self,
-            "Record": Record,
-            "OrderedDict": OrderedDict,
-            "_itemgetter": _itemgetter,
-            "_property": property,
-            "RECORD_VERSION": RECORD_VERSION,
+        {
+            "foo": RecordField("foo", "string"),
+            "bar": RecordField("bar", "varint"),
         }
-        for field in all_fields.values():
-            data["_field_{}".format(field.name)] = field
-
-        exec(c, data)
 
-        self.recordType = data[name.replace("/", "_")]
-
-        self.identifier = (self.name, self.descriptor_hash)
+        Returns:
+            Mapping of Record fields
+        """
+        if self._fields is None:
+            self._fields = OrderedDict([(n, RecordField(n, _type)) for _type, n in self._field_tuples])
+        return self._fields
 
-    @staticmethod
-    def get_required_fields():
+    def get_all_fields(self) -> Mapping[str, RecordField]:
         """
-        Get required fields.
+        Get all fields including required meta fields. eg:
+
+        {
+            "ts": RecordField("ts", "datetime"),
+            "foo": RecordField("foo", "string"),
+            "bar": RecordField("bar", "varint"),
+            "_source": RecordField("_source", "string"),
+            "_classification": RecordField("_classification", "datetime"),
+            "_generated": RecordField("_generated", "datetime"),
+            "_version": RecordField("_version", "vaeint"),
+        }
 
         Returns:
-            OrderedDict
-
+            Mapping of all Record fields
         """
-        required_fields = OrderedDict([(k, RecordField(k, v)) for k, v in RESERVED_FIELDS.items()])
-        return required_fields
+        if self._all_fields is None:
```

Comment:
Is there a noticeable different between doing it like this and decorating this function with `@functools.cache`?

Reply:
This is a class instance method, so it's not recommended to use `@functools.cache`. 

I know you also know this :) but for reference: https://rednafi.github.io/reflections/dont-wrap-instance-methods-with-functoolslru_cache-decorator-in-python.html